### PR TITLE
Implementation for Start and Stop Dialog

### PR DESCRIPTION
**Pull request by ajitht-msft**
# Description

Added implementation for Start and Stop Dialog APIs

Based off of the following swagger spec: https://github.com/williamzhao87/azure-rest-api-specs/blob/dev-communication-CallAutomation-v4-2023-01-15/specification/communication/data-plane/CallAutomation/readme.md

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+     * Instance of CallDialogAsync.
+     */
+    private final CallDialogAsync callDialogAsync;
+
+    /**
+     * Constructor.
+     *
+     * @param callDialogAsync callDialogAsync
+     */
+    public CallDialog(CallDialogAsync callDialogAsync) {
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:29]: don't need public constructor as this is getting retrieved from getCallDialog method from callconnection

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    /**
+     * Start Dialog
+     *
+     * @param options A {@link StartDialogOptions} object containing different options for startDialog.
+     * @throws HttpResponseException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return Response for successful startDialog request.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Response<DialogStateResult> startDialog(StartDialogOptions options) {
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:42]: What is the need of this overload when we already have another method that takes options + operationContext and Context? (line 57)

Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:42]: It is named in a way that this should not return with Response<type> ?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Response<DialogStateResult> startDialogWithResponse(StartDialogOptions options, String operationContext, Context context) {
+        return callDialogAsync.startDialogWithResponseInternal(options, operationContext, context).block();
+    }
+
+    /**
+     * Stop Dialog
+     *
+     * @param dialogId A {@link StartDialogOptions} object containing different options for startDialog.
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:64]: is this accurate description of dialogId?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public void stopDialog(String dialogId) {
+        callDialogAsync.stopDialog(dialogId).block();
+    }
+
+    /**
+     * Stop Dialog
+     *
+     * @param dialogId A {@link StartDialogOptions} object containing different options for startDialog.
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:76]: is this accurate description of dialogId?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * Start Dialog
+     *
+     * @param dialogId startDialog options
+     * @param operationContext operationContext (pass null if not applicable)
+     * @return Response for successful startDialog request.
+     */
+    public Mono<Response<Void>> stopDialogWithResponse(String dialogId, String operationContext) {
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:104]: It seems this stopDialogWithResponse async method have operationContext, but sync stopDialogWithResponse doesnt. I am wondering if this is intended?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    private final String callConnectionId;
+    private final ClientLogger logger;
+
+    /**
+     * Constructor.
+     *
+     * @param callConnectionId callConnectionId the id of the call connection
+     * @param callDialogInternal Internal implementation of CallDialog
+     */
+    public CallDialogAsync(String callConnectionId, CallDialogsImpl callDialogInternal) {
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:38]: don't need public constructor as this is getting retrieved from getCallDialog method from callconnection

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * Start Dialog
+     *
+     * @param startDialogOptions startDialog options
+     * @return Response for successful startDialog request.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<DialogStateResult>> startDialog(StartDialogOptions startDialogOptions) {
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:51]: It is named in a way that this should not return with Response ?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                map(response -> new SimpleResponse<>(response, DialogStateResponseConstructorProxy.create(response.getValue())));
+        } catch (RuntimeException e) {
+            return monoError(logger, e);
+        }
+    }
+
+    /**
+     * Start Dialog
+     *
+     * @param dialogId startDialog options
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:90]: is this the correct param description?

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+            return dialogsInternal.startDialogWithResponseAsync(callConnectionId, startDialogOptions.getDialogId(), requestInternal, context).
+                map(response -> new SimpleResponse<>(response, DialogStateResponseConstructorProxy.create(response.getValue())));
+        } catch (RuntimeException e) {
+            return monoError(logger, e);
+        }
+    }
+
+    /**
+     * Start Dialog
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:88]: it says startDialog in summary but the method is called stopDialog

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @param dialogId startDialog options
+     * @return Response for successful startDialog request.
+     */
+    public Mono<Void> stopDialog(String dialogId) {
+        return stopDialogWithResponse(dialogId, null).then();
+    }
+
+    /**
+     * Start Dialog
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:98]: it says startDialog in summary but the method is called stopDialog

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+     *
+     * @param dialogId startDialog options
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:100]: same param description discrepancy here

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+     * @param operationContext operationContext (pass null if not applicable)
+     * @return Response for successful startDialog request.
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialogAsync.java:102]: stopDialog

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallConnection.java [hunk 13]
```diff
@@ -274,5 +274,15 @@ public CallMedia getCallMedia() {
         return new CallMedia(callConnectionAsync.getCallMediaAsync());
     }
 
+    /***
+     * Returns an object of CallContent
```
Comment (minwoolee-msft) [inline sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallConnection.java:278]: Is this the correct description?

**Review by juntuchen-msft — CHANGES_REQUESTED** [review]

File: sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java (hunk 2 above)
Comment (ajitht-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:42]: i was simply following the structure of the dtmf actions as a blueprint. But if this is not necessary i will remove.

Comment (ajitht-msft) [reply sdk/communication/azure-communication-callautomation/src/main/java/com/azure/communication/callautomation/CallDialog.java:42]: removing this method altogether.

**Review by juntuchen-msft — APPROVED** [review]

**Comment by minwoolee-msft** [comment]
It seems that following sync methods are deleted: 
> startDialog
> stopDialog

This should be present.

The comment i left above regarding this is about its signature. Look at the following: 
> public Mono<Response<DialogStateResult>> startDialog(StartDialogOptions startDialogOptions) 
> public Mono<Response<DialogStateResult>> startDialogWithResponse(StartDialogOptions startDialogOptions, String **operationContext**)

From above, why is operationContext be part of signature? Options pattern should include this.
See the guideline on this here: https://azure.github.io/azure-sdk/java_introduction.html#option-parameters

**Review by juntuchen-msft — APPROVED** [review]

**Review by minwoolee-msft — APPROVED** [review]

**Pull request merged**